src/track/beats.h

```diff
@@ -81,8 +96,30 @@ class Beats : private std::enable_shared_from_this<Beats> {
             updateValue();
         }
 
+        /// Returns the frame duration between this beat and the next beat.
         mixxx::audio::FrameDiff_t beatLengthFrames() const;
 
+        /// Returns the number of beats per bar.
+        ///
+        /// Note: Currently always returns 4, as time signatures other than 4/4
+        /// are not supported yet.
+        int beatsPerBar() const;
+
+        /// Returns the current index of the beat in the bar.
+        int beatIndexInBar() const {
+            return m_beatOffset % beatsPerBar();
```

Comment:
This assumes that the markers always align with a bar. This is probably musically correct. 
However the model allows overlapping bars, and especially when we have dynamic beat-gits created in earlier version of Mixxx, we can have tempo changes in any place. We need to deal with it. 

From the GUI perceptive, I would be in favour to require a bar for a temp change.

Reply:
for cases where the tempo is increasing gradually, every beat may be a little faster or slower than the last.  We should allow tempo changes on any beat marker.

Reply:
Perhaps a second parameter would be useful to describe accelerando/ritardando in beats/s². On the other hand, the case might be too rare to justify this overhead.

Reply:
I agree with @ywwg, tempo changes on every beat are common enough that we should support them. The problem I see is that it's probably not easy to retrofit into our current beatgrid model (that has no information about bars/measures) if we cannot interpret markers as downbeats. So we'd probably have to defer this for now, just like changing time signatures, which I'd very much be interested in having too.

Reply:
I see a chance to integrate that. 
For my understanding we need to add a an bar offset along with the beats per bar value. 
This way we are still able to express the old free beat map with the new model

```
     ‖___|___|___|___‖___|__|__|__‖__|__|___|___‖___     
     M                  M              M  . 
     120bpm.            180 bpm        120 bpm   
     barAt 0.           barAt 3        barAt 2
     beatsPerBar 4      beatsPerBar 4  beatsPerBar 4
 ```

Reply:
it may well be that bpm-per-measure is "close enough" that it works with bpm speed-up/slowdowns or non-quantized tracks (disco etc). If we can test this situation early then we'll know if it's a problem or not.  I have a few tracks that are my go-tos for difficult beatgrids, I can send them to you @fwcd if that helps